Sources/LanguageServerProtocol/LanguageServer.swift

```diff
@@ -10,185 +10,207 @@
+            }
+      }
+      open func _logNotification<N>(_ notification: Notification<N>) {
+            logAsync { _ in
+                  "\(type(of: self)): \(notification)"
+            }
+      }
+      open func _logResponse<Response>(
+            _ result: LSPResult<Response>, id: RequestID, method: String
+      ) {
+            logAsync { _ in
+                  """
       \(type(of: self)): Response<\(method)>(
         \(result)
       )
       """
-    }
-  }
+            }
+      }
 }
 
 extension LanguageServerEndpoint {
-  // MARK: Request registration.
-
-  /// Register the given request handler, which must be a method on `self`.
-  ///
-  /// Must be called on `queue`.
-  public func _register<Server, R>(_ requestHandler: @escaping (Server) -> (Request<R>) -> Void) {
-    // We can use `unowned` here because the handler is run synchronously on `queue`.
-    precondition(self is Server)
-    requestHandlers[ObjectIdentifier(R.self)] = { [unowned self] request in
-      requestHandler(self as! Server)(request)
-    }
-  }
-
-  /// Register the given notification handler, which must be a method on `self`.
-  ///
-  /// Must be called on `queue`.
-  public func _register<Server, N>(_ noteHandler: @escaping (Server) -> (Notification<N>) -> Void) {
-    // We can use `unowned` here because the handler is run synchronously on `queue`.
-    notificationHandlers[ObjectIdentifier(N.self)] = { [unowned self] note in
-      noteHandler(self as! Server)(note)
-    }
-  }
-
-  /// Register the given request handler.
-  ///
-  /// Must be called on `queue`.
-  public func _register<R>(_ requestHandler: @escaping (Request<R>) -> Void) {
-    requestHandlers[ObjectIdentifier(R.self)] = requestHandler
-  }
-
-  /// Register the given notification handler.
-  ///
-  /// Must be called on `queue`.
-  public func _register<N>(_ noteHandler: @escaping (Notification<N>) -> Void) {
-    notificationHandlers[ObjectIdentifier(N.self)] = noteHandler
-  }
-
-  /// Register the given request handler.  **For test messages only**.
-  public func register<R>(_ requestHandler: @escaping (Request<R>) -> Void) {
-    queue.sync { _register(requestHandler) }
-  }
-
-  /// Register the given notification handler.  **For test messages only**.
-  public func register<N>(_ noteHandler: @escaping (Notification<N>) -> Void) {
-    queue.sync { _register(noteHandler) }
-  }
+      // MARK: Request registration.
+
+      /// Register the given request handler, which must be a method on `self`.
+      ///
+      /// Must be called on `queue`.
+      public func _register<Server, R>(
+            _ requestHandler: @escaping (Server) -> (Request<R>) -> Void
+      ) {
+            // We can use `unowned` here because the handler is run synchronously on `queue`.
+            precondition(self is Server)
+            requestHandlers[ObjectIdentifier(R.self)] = {
+                  [unowned self] request in
+                  requestHandler(self as! Server)(request)
```

Comment:
Having the capture list and signature on the same indentation as the body seems like a mistake.

Reply:
That's intended.  How would you indent it?

Reply:
It's hopefully a rare case, but I prefer keeping the signature separate, so add an additional level of indent:

```
requestHandlers[ObjectIdentifier(R.self)] = {
                                    [unowned self] request in
                  requestHandler(self as! Server)(request)
```

Reply:
Closures are one of the two biggest pain points we've had to work with when trying to universally apply the idea of separating the body of a block from the prologue above it (`case` clauses are the other one). We've yet to find an answer that we _loved_ here.

Google's original style guide suggested that if the signature of a closure must be moved down to its own line, it should be surrounded by parentheses to help distinguish it (albeit slightly) from the rest of the body, but inserting or removing non-whitespace characters as part of the line-wrapping algorithm starts getting a little weird.

We could also just force a blank line between them, but I'm not crazy about that either.

Reply:
My take is that 

- yes, it is rare
- the alternative proposed is quirky—no other rule increases indentation by more than 1 between consecutive lines.
- the programmer can always add a blank line after the "in" clause
- therefore, it's not worth complicating the rules to do something special here.